### PR TITLE
mtd-recovery: Add a plugin that knows how to MUX and load an MTD device

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [X] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/mtd-recovery/README.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+This plugin supports the following protocol ID:
+
+* org.infradead.mtd
+
+## GUID Generation
+
+These devices are generated from the baseboard vendor ID:
+
+* `MTD\VEN_Lenovo&DEV_ThinkPad`
+* `MTD\VEN_Lenovo`
```
Comment (superm1) [inline plugins/mtd-recovery/README.md:16]: If the most common case is from behind a BMC then I think this is very unlikely to ever show up in a laptop...  Maybe a different example from a server makes sense.

File: plugins/mtd-recovery/fu-mtd-recovery-device.c [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+{
+	FuMtdRecoveryDevice *self = FU_MTD_RECOVERY_DEVICE(device);
+	g_autoptr(FuUdevDevice) device_tmp = NULL;
+
+	/* already available */
+	if (fu_device_get_proxy(device) != NULL)
+		return TRUE;
+
+	/* create GPIO device and disable write lock */
+	if (self->gpio_number != 0) {
```
Comment (superm1) [inline plugins/mtd-recovery/fu-mtd-recovery-device.c:133]: why can't it be GPIO 0?

Comment (hughsie) [reply plugins/mtd-recovery/fu-mtd-recovery-device.c:133]: Agreed, fixed.

**Comment by hughsie** [comment]
> Is this to enable a "new" design or an "existing" design?

In this case, an existing, but non-public design deployed all over the place.

>  it's begging a spec to specify in the ACPI/DT tables that a certain GPIO is connected to a mux that can do this.

Yes, this is what I'm lobbying for too. Another security company wants to know the same information and it's pretty crap that we'd have to hardcode it per-model, perhaps in more than one project.

> presumably other things can potentially be exposed from a mux controlled by GPIO?

Hmm, perhaps -- you mean like a USB device?

> Swapping the GPIO to expose the MTD device has to have some implication for whatever is not connected anymore...

Agreed, also a concern. Apparently the "host OS" gets writes denied, i.e. a kernel write error rather than anything electrically bad. I believe in the demo hardware that we were developing against the host OS has to be powered down, although I'm not sure of the mechanism that enforces this.

**Comment by superm1** [comment]
> have to hardcode it per-model, perhaps in more than one project.

What other projects?  If it's coreboot or an open source BSP firmware then it would be better to potentially introduce the mechanism in that other project to introspect what GPIO it is (advertise to a table, a mailbox message, _anything really_)

> Hmm, perhaps -- you mean like a USB device?

Yeah I mean I don't know the details of what actually happens when you trigger this GPIO.  If it's really a mux then it has something connected when set one way and something else when connected another way.  If it's not really a mux, it might be toggling a write protect pin for something.

> Apparently the "host OS" gets writes denied, i.e. a kernel write error rather than anything electrically bad

Which jives more to the GPIO being on a _write protect pin not a mux_. 

> host OS has to be powered down, although I'm not sure of the mechanism that enforces this.

Eek!

**Comment by hughsie** [comment]
> What other projects?

In this example it'd be some kind of security firmware agent, so not the kind of thing we expect to be installed by default. I think this kind of data really belongs in DT or ACPI somewhere. That's what I'm advocating for future hardware.

>  open source BSP firmware

I'll investigate what OpenBMC could provide; the medium term plan is for OpenBMC to *use* fwupd to update stuff, so I'm kinda expecting *not our problem, it's yours!* :)

> If it's really a mux then it has something connected when set one way and something else when connected another way. 

This is what I was lead to believe.

> it might be toggling a write protect pin for something.

Nah, without the GPIO set the kernel driver just refuses to load, so I actually think it's controlling the clock and data.

> Eek!

Yes, I should dive more into this -- e.g. setting an inhibit on the device if the host is powered on. That might be phase #2.

**Comment by superm1** [comment]
> Nah, without the GPIO set the kernel driver just refuses to load, so I actually think it's controlling the clock and data.

I bet it's a mux that pulls the clock into a pull down or clock to a real clock pin then.  

> Yes, I should dive more into this -- e.g. setting an inhibit on the device if the host is powered on. That might be phase 2

Yeah, I hope that there can be an indication from the BMC to make this design safe.

File: plugins/mtd-recovery/README.md (hunk 1 above)
Comment (hughsie) [reply plugins/mtd-recovery/README.md:16]: Yup, changed.

**Comment by hughsie** [comment]
Lets close this PR but keep the branch -- if the vendor has interest in doing this we can ask them to make it "just work" at the same time.

**Pull request closed without merging**